src/hardware/input/mouseif_ps2_bios.cpp

```diff
@@ -217,50 +294,155 @@ void MOUSEPS2_UpdatePacket()
 		                static_cast<int16_t>(-UINT8_MAX),
 		                static_cast<int16_t>(UINT8_MAX));
 	} else {
-		if ((dx > UINT8_MAX) || (dx < -UINT8_MAX))
+		if ((dx > UINT8_MAX) || (dx < -UINT8_MAX)) {
 			mdat.overflow_x = 1;
-		if ((dy > UINT8_MAX) || (dy < -UINT8_MAX))
+		}
+		if ((dy > UINT8_MAX) || (dy < -UINT8_MAX)) {
 			mdat.overflow_y = 1;
+		}
 	}
 
 	dx = static_cast<int16_t>(dx % (UINT8_MAX + 1));
 	if (dx < 0) {
-		dx          = static_cast<int16_t>(dx + UINT8_MAX + 1);
+		dx = static_cast<int16_t>(dx + UINT8_MAX + 1);
 		mdat.sign_x = 1;
 	}
 
 	dy = static_cast<int16_t>(dy % (UINT8_MAX + 1));
 	if (dy < 0) {
-		dy          = static_cast<int16_t>(dy + UINT8_MAX + 1);
+		dy = static_cast<int16_t>(dy + UINT8_MAX + 1);
 		mdat.sign_y = 1;
 	}
 
-	packet[0] = mdat.data;
-	packet[1] = static_cast<uint8_t>(dx);
-	packet[2] = static_cast<uint8_t>(dy);
+	frame.resize(3);
+	frame[0] = mdat.data;
+	frame[1] = static_cast<uint8_t>(dx);
+	frame[2] = static_cast<uint8_t>(dy);
+
+	if (protocol == MouseModelPS2::IntelliMouse) {
+		frame.resize(4);
+		frame[3] = get_reset_wheel_8bit();
+	} else if (protocol == MouseModelPS2::Explorer) {
+		frame.resize(4);
+		frame[3] = get_reset_wheel_4bit();
+		if (buttons.extra_1) {
+			bit::set(frame[3], b4);
+		}
+		if (buttons.extra_2) {
+			bit::set(frame[3], b5);
+		}
+	}
 
-	if (protocol == MouseModelPS2::IntelliMouse)
-		packet[3] = get_reset_wheel_8bit();
-	else if (protocol == MouseModelPS2::Explorer) {
-		packet[3] = get_reset_wheel_4bit();
-		if (buttons.extra_1)
-			bit::set(packet[3], b4);
-		if (buttons.extra_2)
-			bit::set(packet[3], b5);
-	} else
-		packet[3] = 0;
+	// Protocol frame was build, no need for a second one
+	has_data_for_frame = false;
+}
+
+static void maybe_transfer_frame(); // forward declaration
+
+static void delay_handler(uint32_t /*val*/)
+{
+	delay_running  = false;
+	delay_finished = true;
+
+	maybe_transfer_frame();
+}
+
+static void maybe_start_delay_timer(const uint8_t timer_delay_ms)
+{
+	if (delay_running) {
+		return;
+	}
+	PIC_AddEvent(delay_handler, timer_delay_ms);
+	delay_running  = true;
+	delay_finished = false;
+}
+
+static bool should_report()
+{
+	return !mode_wrap && !mode_remote && is_reporting;
+}
+
+static void maybe_transfer_frame()
+{
+	if (!delay_finished) {
+		maybe_start_delay_timer(delay_ms);
+		return;
+	}
+
+	if (!has_data_for_frame || !I8042_IsReadyForAuxFrame()) {
+		return;
+	}
+
+	maybe_start_delay_timer(delay_ms);
+
+	build_protocol_frame();
+	if (should_report()) {
+		I8042_AddAuxFrame(frame);
+	}
+
+	frame.clear();
+	return;
```

Comment:
No need for explicit return at the end of function.

Reply:
OK, removed.